routes/Authentication.js

```diff
@@ -88,55 +86,49 @@ router.post("/facebook/login", passport.authenticate("facebookToken"), function(
-      });
-    } else {
-      var hashedPassword = bcrypt.hashSync(req.body.password, 8);
-      //Create token for email verification
-      const secretToken = randomString.generate();
-
-      User.create(
-        {
-          name: req.body.name,
-          email: req.body.email,
-          password: hashedPassword,
-          //For email verification
-          secretToken: secretToken,
-          isActivated: false
-        },
-        function(err, user) {
-          if (err)
-            return res
-              .status(500)
-              .send("There was a problem registering the user.");
-          // create a token
-          //var token = jwt.sign({ id: user._id }, config.secret, {
-          //  expiresIn: 86400 // expires in 24 hours
-          //});
-          res.status(200).send(user.email);
-          const mailOptions = {
-            from: "support@mirrorwall.org", // sender address
-            to: user.email, // list of receivers
-            subject: "Please verify the email address", // Subject line
-            html: `<h4>Thank you for registering.<h4>
-            <p>Please verify this email address by clicking the following link</p>
-            <a href="https://frozen-lake-54898.herokuapp.com/api/auth/verify/${
-              user.secretToken
-            }">
-            https://frozen-lake-54898.herokuapp.com/api/auth/verify/${
-              user.secretToken
-            }</a>`
-          };
-
-          mailer.sendMail(mailOptions);
-        }
-      );
+      return res
+        .status(409)
+        .json({ message: "This e-mail is already registered" });
     }
-  });
+    let hashedPassword = bcrypt.hashSync(req.body.password, 8);
+    const secretToken = randomString.generate();
+    let newUser = await User.create({
+      name: req.body.name,
+      email: req.body.email,
+      password: hashedPassword,
+      //For email verification
+      secretToken: secretToken,
+      isActivated: false
+    });
+    if (newUser) {
+      console.log(newUser);
+      const mailOptions = {
+        from: "support@mirrorwall.org", //sender address
+        to: newUser.email, //list of receivers
+        subject: "Please verify the email address", // Subject line
+        html: `<h4>Thank you for registering.<h4>
+        <p>Please verify this email address by clicking the following link</p>
+        <a href="https://frozen-lake-54898.herokuapp.com/api/auth/verify/${newUser.secretToken}">
+        https://frozen-lake-54898.herokuapp.com/api/auth/verify/${newUser.secretToken}</a>`
+      };
+      let mailResponse = await mailer.sendMail(mailOptions);
+      if (mailResponse.messageId) {
+        res.status(200).send(newUser.email);
+      } else {
+        let deleted = await User.deleteOne({ email: req.body.email });
+        console.log(deleted);
+        throw new Error("Something wrong");
+      }
+    }
+  } catch (err) {
+    console.log(err);
+    let deleted = await User.deleteOne({ email: req.body.email });
```

Comment:
Hi, would you please explain why this (deletion) is necessary? Thanks :)

Reply:
As i think if the email is not sent to the user registering then we should delete the user from the db and let him register again(giving error in the frontend as Something went wrong please register again).

Reply:
Yes, That's reasonable if the failure occurred due to a wrong email address. But it also can fail due to an error in server, in which case its unfair to ask user to enter details again. But, I'm merging this because its still a better approach than the existing. Thanks.